File: TS_JT-Reader/TS_JT-Reader_Vis-Studio/TS_JT-Reader/BaseJTNode.ts
module C_sharp_JT_Reader {
    export class BaseJTNode {
        protected _uncompressed: number[];
        protected _fileVersion: number = 0;
        protected _richTextBox: RichTextBox;
        protected _filePosCount: number = 0;
        protected fileBytes: number[] = new Array(4);
        protected _nodeFlags: UInt32;
        protected _attributeCount: Int32;
        protected _attributeObjectID: Int32[];
        protected _objectID: Int32;
        protected _versionNumber: Int16;
        protected m_textBox: List<string> = new List<string>();
        constructor() {

        }
        constructor(name: string, type: string) {

        }
        protected TraverseBaseNodeData(): number {
            if (this._fileVersion >= 9.5) {
                Buffer.BlockCopy(this._uncompressed, this._filePosCount, this.fileBytes, 0, 4);
                this._versionNumber = DataTypes.getInt16(this.fileBytes);
                this.m_textBox.Add("\nVersion number = " + this._versionNumber.ToString());
                this._filePosCount += 2;
            }
            else {
                Buffer.BlockCopy(this._uncompressed, this._filePosCount, this.fileBytes, 0, 4);
                this._objectID = DataTypes.getInt32(this.fileBytes);
                this.m_textBox.Add("\nObject ID = " + this._objectID.ToString());
                this._filePosCount += 4;
            }
            Buffer.BlockCopy(this._uncompressed, this._filePosCount, this.fileBytes, 0, 4);
            this._nodeFlags = DataTypes.getUInt32(this.fileBytes);
            var a: number = this.fileBytes[0];
            var b: number = this.fileBytes[2];
            var c: number = <number>(b | a);
            this.m_textBox.Add("\nNode Flags = " + this._nodeFlags.ToString());
            this._filePosCount += 4;
            if (this._nodeFlags == 0) {
                Buffer.BlockCopy(this._uncompressed, this._filePosCount, this.fileBytes, 0, 4);
                this._attributeCount = DataTypes.getInt32(this.fileBytes);
                this.m_textBox.Add("\nAttribute Count = " + this._attributeCount.ToString());
                this._filePosCount += 4;
                this._attributeObjectID = new Array(this._attributeCount);
                for (var i: number = 0; i < this._attributeCount; i++) {
                    Buffer.BlockCopy(this._uncompressed, this._filePosCount, this.fileBytes, 0, 4);
                    this._attributeObjectID[i] = DataTypes.getInt32(this.fileBytes);
                    this.m_textBox.Add("\nAttribute Object ID = " + this._attributeObjectID[i].ToString());
                    this._filePosCount += 4;
                }
            }
            else {
                this._filePosCount += 4 + 4;
            }
            return this._filePosCount;
        }
    }
}

File: TS_JT-Reader/C#_JT-Reader_Vis-Studio/C sharp JT-Reader/Nodes/BaseJTNode.cs
/////////////////////////////////////////////////////////////////////
//
// This class contains the Base Node Data used by the JT File
//
// Refer to ????????   7.2.6 Meta Data Segment page 162 - on (Ver 9.5 rev D)
//
/////////////////////////////////////////////////////////////////////


using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Drawing;
using System.Windows.Forms;

namespace C_sharp_JT_Reader
{
    public class BaseJTNode
    {
        protected byte[] _uncompressed;
        protected float _fileVersion = 0;
        protected RichTextBox _richTextBox;
        protected int _filePosCount = 0;
        protected byte[] fileBytes = new byte[4];
        protected UInt32 _nodeFlags;
        protected Int32 _attributeCount;
        protected Int32[] _attributeObjectID;
        protected Int32 _objectID;
        protected Int16 _versionNumber;

        protected List<string> m_textBox = new List<string>();


        protected BaseJTNode()
        {

        }

        protected BaseJTNode(string name, string type)
        {

        }

        protected int TraverseBaseNodeData()
        {
            if (_fileVersion >= 9.5)
            {
                Buffer.BlockCopy(_uncompressed, _filePosCount, fileBytes, 0, 4);
                _versionNumber = DataTypes.getInt16(fileBytes);
                this.m_textBox.Add("\nVersion number = " + _versionNumber.ToString());
                _filePosCount += sizeof(Int16);
            }
            else
            {
                Buffer.BlockCopy(_uncompressed, _filePosCount, fileBytes, 0, 4);
                _objectID = DataTypes.getInt32(fileBytes);
                this.m_textBox.Add("\nObject ID = " + _objectID.ToString());
                _filePosCount += sizeof(Int32);
            }

            Buffer.BlockCopy(_uncompressed, _filePosCount, fileBytes, 0, 4);
            _nodeFlags = DataTypes.getUInt32(fileBytes);
            byte a = fileBytes[0];
            byte b = fileBytes[2];
            byte c = (byte)(b|a);//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

            this.m_textBox.Add("\nNode Flags = " + _nodeFlags.ToString());
            _filePosCount += sizeof(UInt32);

            if (_nodeFlags == 0)
            {
                Buffer.BlockCopy(_uncompressed, _filePosCount, fileBytes, 0, 4);
                _attributeCount = DataTypes.getInt32(fileBytes);
                this.m_textBox.Add("\nAttribute Count = " + _attributeCount.ToString());
                _filePosCount += sizeof(Int32);

                _attributeObjectID = new Int32[_attributeCount];
                for (int i = 0; i < _attributeCount; i++)
                {
                    Buffer.BlockCopy(_uncompressed, _filePosCount, fileBytes, 0, 4);
                    _attributeObjectID[i] = DataTypes.getInt32(fileBytes);
                    this.m_textBox.Add("\nAttribute Object ID = " + _attributeObjectID[i].ToString());
                    _filePosCount += sizeof(Int32);
                }
            }
            else
            {
                _filePosCount += sizeof(Int32) + sizeof(Int32);
            }
            return _filePosCount;
        }

    }
 }
